### ISSUE TITLE
Interleave test items in dvsim.py runs

### ISSUE THREAD

**Issue by rswarbrick** [lowRISC/opentitan#2370]
If I've made a change to a testbench (or the design!), I'm going to want to run some tests to make sure I haven't broken it. So I'll run something like

    path/to/dvsim.py some-cfg.hjson --reseed 50

Suppose that `some-cfg.hjson` defines 10 tests. Then this is going to be 500 simulations, which might take a while. At the moment, we run 50 iterations of test 1, then 50 iterations of test 2, etc.

If I've broken something, it's most likely to only break some tests. If I'm unlucky, I might have just broken test 10. Oh no! It would be more helpful when running at the terminal if the tests were interleaved.

**Comment by rswarbrick** [comment]
@sriyerg: Does this seem reasonable to you? If so, I don't think it would be a difficult change in `Deploy.py` and I'll happily make the patch, but I wanted to run the idea past someone before doing the work.

**Comment by rswarbrick** [comment]
Thinking about this, we might not be forcing all tests to run the same number of iterations with `--reseed`. Maybe it's most helpful just to randomise the dispatch queue. The same tests will be run as now, but they will finish in a (hopefully) more helpful order.

**Comment by sriyerg** [comment]
If by interleave, you mean 'dispatch the iteration 1 of all tests, followed by iteration 2 ...', as opposed to 'dispatch all iterations of test 1, followed by all iterations of test 2 ...', then I agree - it sounds like a good idea. I prefer it over a randomized queue. 

I think the seed / reseed part need not be touched - IMO it is already WAI. Just need to reorder the `dispatch_items`.

**Comment by rswarbrick** [comment]
I think that we have to be slightly more careful than "iteration 1 of everything; iteration 2 of everything". Suppose that there are 2 tests, A and B where A takes 10 times as long as B. Then the default configuration might run A 10 times and B 100 times. Ideally, we'd run ABBBBBBBBBBABBBBBBBBBB etc. Doing that right in general looked like it would take some thought, which is why I suggested randomising the order (which will usually approximate the interleaved schedule), but I think I can see how to do it properly. I'll try to push up a patch in the next couple of days.

**Comment by sriyerg** [comment]
Typically, on a change we want to see the full suite of tests run and pass at least for one iteration. So if A runs 5 times, B runs 10 times, it would look like this: ABABABABABBBBBB. It doesn't matter if A runs N times as long as B, we just need to ensure we prioritize on kicking off all unique tests first, so that we can get the feedback of the changes we made deterministically early.

**Comment by rswarbrick** [comment]
I agree, and the PR I pushed achieves that. The 1st run for each test gets a "tag" of zero, so the sequence will look something like `ABCBABBABCBABBABB`. Note that the sequence starts with `ABC`. Maybe it would be better to have further discussion on the PR.

**Comment by rasmus-madsen** [comment]
@rswarbrick one question.
if two tests have different build time dependencies - then you would need recompilation in between wouldn't you?

**Comment by rswarbrick** [comment]
In the dvsim model, there is a single "build" phase that runs VCS/xcelium/whatever to build one or more tests (specified by test class & top-level sequence). Then there's a "run" phase, that runs the these tests with one or more random seeds.

Here, I'm talking about the order that we should run stuff that was generated by a single build, and arguing that we should interleave runs for different tests (i.e. test class/top-level sequence) rather than running all seeds for one test, then all seeds for the next etc.

PS: Gah! I wish there were more nouns available. It feels like "test" means way too many things.

**Comment by rasmus-madsen** [comment]
@rswarbrick I agree TEST is used to widely here.
I think we could all be better as using the correct names.
what is typically called test in OpenTitan should be referred to as test sequence IMO.

**Comment by sriyerg** [comment]
> @rswarbrick one question.
> if two tests have different build time dependencies - then you would need recompilation in between wouldn't you?

Build time dependencies are set using `build_mode` key. If this key is not present, it uses the `default` build mode (which is the 'vanilla' build without any extra options). If a test needs specific build options, it needs to wrap it in a unique `build_mode`. See example below:

```
  // in aes_sim_cfg.hjson, for example
  build_modes: [
    {
      name: aes_foo
      // Set build-time switches
      build_opts: ["+define+foo=1", "+define+bar=2"]
      // Build mode can enable run-time switches as well
      run_opts: ["+foo=1", "+bar=2"]
    }
  ]

  tests: [
    ...
    {
      name: aes_foo_test
      uvm_test: aes_foo_test
      uvm_seq: aes_test_vseq
      build_mode: aes_foo 
    }
  ]
```

When you specify tests that need to be run, `dvsim` will figure out how many builds are to be dispatched. Example: if you specify on the command line `-i aes_sanity_test aes_foo_test`, it will dispatch 2 builds in parallel, `default` and `aes_foo` first. As each build finishes, it will dispatch the runs that cater to it. So recompilation is not needed (since the compilation happens in parallel).

**Comment by rswarbrick** [comment]
Fixed by PR #2400.
